### PR TITLE
Add exemplar support to OpenCensus bridge

### PR DESCRIPTION
**Pull request by dashpole**
Fixes https://github.com/open-telemetry/opentelemetry-go/issues/4572

### CODE REVIEW

**Comment by dashpole** [comment]
I've split out the error handling changes into https://github.com/open-telemetry/opentelemetry-go/pull/4599.  I'll wait to mark it ready for review until it is rebased

File: bridge/opencensus/internal/ocmetric/metric.go
Comment (pellared) [inline bridge/opencensus/internal/ocmetric/metric.go:185]: I think that the following would be more idiomatic (mostly https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow).

I also think that converting exemplars is not needed when `bucket.Count < 0` as the caller ignores the result if there is an error anyway.

```suggestion
			err = errors.Join(err, fmt.Errorf("%w: %q", errNegativeBucketCount, bucket.Count))
			continue
		}
		bucketCounts[i] = uint64(bucket.Count)

		if bucket.Exemplar != nil {
			exemplar, exemplarErr := convertExemplar(bucket.Exemplar)
			if exemplarErr != nil {
				err = errors.Join(err, exemplarErr)
				continue
			}
			exemplars = append(exemplars, exemplar)
		}
```

**Review by pellared — COMMENTED** [review]
Nice 👍 I left some minor comments

File: bridge/opencensus/internal/ocmetric/metric.go
Comment (MrAlias) [inline bridge/opencensus/internal/ocmetric/metric.go:214]: The OTel filtered attributes are intended to be the attributes not included in metric data point the exemplar is associated with. Is this the same for OpenCensus?

**Review by pellared — APPROVED** [review]

File: bridge/opencensus/internal/ocmetric/metric.go
Comment (MrAlias) [inline bridge/opencensus/internal/ocmetric/metric.go:224]: This is missing cases we could handle:

- `int8` -> `Int`
- `int16` -> `Int`
- `int32` -> `Int`
- `uint`  -> `Int` or `String` depending on size
- `uint8` -> `Int`
- `uint16` -> `Int`
- `uint32` -> `Int`
- `uint64` -> `Int` or `String` depending on size
- `uintptr` -> `Int` or `String` depending on size
- `float32` -> `Float64`
- `complex64` -> `String`
- `complex128` -> `String`

It might also make sense to include a generic "unhandled" attribute that will contain all unhanded values so the user knows about them instead of dropping.

I did something similar [here](https://github.com/MrAlias/otlpr/blob/081674469a3c0ce835143473668c235569aaea7a/internal/formatter.go#L132-L293) for the conversion to OTLP directly. It's not a direct comparison, but maybe it could be a helpful reference.

File: bridge/opencensus/internal/ocmetric/metric.go
Comment (dashpole) [reply bridge/opencensus/internal/ocmetric/metric.go:214]: TBH, i'm not sure.  Attachments can be added to exemplars using the [WithAttachments](https://github.com/census-instrumentation/opencensus-go/blob/01e6da5fc01c42aca1e0ce315f41744876e9fcbb/stats/record.go#L50) option when making an observation.  The only usage of it I can find are in the [OpenCensus grpc instrumentation](https://github.com/census-instrumentation/opencensus-go/blob/01e6da5fc01c42aca1e0ce315f41744876e9fcbb/plugin/ocgrpc/stats_common.go#L173), but that doesn't set anything except span context.  The spec doesn't say what attachments are for: https://github.com/census-instrumentation/opencensus-specs/blob/master/stats/Exemplars.md#attachment, just that they are key-value pairs.

Comment (MrAlias) [reply bridge/opencensus/internal/ocmetric/metric.go:214]: Gotcha. Maybe we just include this as is now and if we discover a problem in the future we can filter out the attributes from the OpenCensus time-series?

File: bridge/opencensus/internal/ocmetric/metric.go
Comment (dashpole) [reply bridge/opencensus/internal/ocmetric/metric.go:219]: done

File: bridge/opencensus/internal/ocmetric/metric.go
Comment (dashpole) [reply bridge/opencensus/internal/ocmetric/metric.go]: Sounds good to me.

**Review by MrAlias — APPROVED** [review]
:rocket:

**Comment by dashpole** [comment]
Hold on.  Looks like I missed test cases for float32...

**Comment by MrAlias** [comment]
@open-telemetry/go-approvers PTAL. This needs one more review from a non-Splunker.

File: bridge/opencensus/internal/ocmetric/metric.go
Comment (MadVikingGod) [inline bridge/opencensus/internal/ocmetric/metric.go:171]: Should the Examplars be the same size as bucketCounts?

Comment (dashpole) [reply bridge/opencensus/internal/ocmetric/metric.go:171]: The OTel [data model](https://github.com/open-telemetry/opentelemetry-proto/blob/ac3242b03157295e4ee9e616af53b81517b06559/opentelemetry/proto/metrics/v1/metrics.proto#L453) doesn't requires the number or order of exemplars to be the same as the set of histogram buckets counts, which must be ordered, and have a matching lenth.

Even it OTel, you would only get one exemplar per-bucket on histograms if you are using the AlignedHistogramBucketExemplarReservoir.  Otherwise, I believe you could end up with multiple exemplars per-bucket, or exemplars that are not aligned with buckets.

OpenCensus does limit each bucket to at most one exemplar, but I don't think we should insert empty exemplars for buckets that don't have one.

**Review by MadVikingGod — APPROVED** [review]

**Pull request merged**
